File: beam-android/core/testing/src/main/kotlin/dev/beam/android/core/testing/Fixtures.kt
package dev.beam.android.core.testing

import uniffi.beam_client_core.AdminCounts
import uniffi.beam_client_core.AdminStatus
import uniffi.beam_client_core.AudioTrackView
import uniffi.beam_client_core.ContinueWatchingEntry
import uniffi.beam_client_core.EnrichmentCounts
import uniffi.beam_client_core.EpisodeSummary
import uniffi.beam_client_core.HistoryEntry
import uniffi.beam_client_core.LibrarySummary
import uniffi.beam_client_core.MediaDetail
import uniffi.beam_client_core.MediaKind
import uniffi.beam_client_core.MediaPage
import uniffi.beam_client_core.MediaSourceView
import uniffi.beam_client_core.MediaSummary
import uniffi.beam_client_core.SeasonSummary
import uniffi.beam_client_core.ServerSummary
import uniffi.beam_client_core.SessionState
import uniffi.beam_client_core.UserSummary

/**
 * Builders for the core's records.
 *
 * Every parameter has a default that produces a valid value, so a test names
 * only the field it is actually about. A test that reads
 * `movie(title = "Le Samourai")` says what it is testing; one that spells out
 * fourteen fields buries it.
 */
public object Fixtures {
    public fun movie(
        id: String = "movie-1",
        title: String = "Le Samourai",
        year: UInt? = 1967u,
        posterUrl: String? = "https://beam.test/poster.jpg",
        fileId: String? = "file-1",
        genres: List<String> = listOf("Crime", "Drama"),
        runtimeMinutes: UInt? = 105u,
        tmdbRating: UInt? = 81u,
        description: String? = "A contract killer plans his last job.",
    ): MediaSummary =
        MediaSummary(
            id = id,
            kind = MediaKind.MOVIE,
            title = title,
            originalTitle = title,
            year = year,
            description = description,
            posterUrl = posterUrl,
            backdropUrl = null,
            genres = genres,
            runtimeMinutes = runtimeMinutes,
            tmdbRating = tmdbRating,
            fileId = fileId,
            seasonCount = 0u,
            episodeCount = 0u,
        )

    public fun show(
        id: String = "show-1",
        title: String = "Le Bureau",
        year: UInt? = 2015u,
        seasonCount: UInt = 2u,
        episodeCount: UInt = 12u,
        posterUrl: String? = "https://beam.test/show.jpg",
        genres: List<String> = listOf("Thriller"),
    ): MediaSummary =
        MediaSummary(
            id = id,
            kind = MediaKind.SHOW,
            title = title,
            originalTitle = title,
            year = year,
            description = "Undercover officers, handled from Paris.",
            posterUrl = posterUrl,
            backdropUrl = null,
            genres = genres,
            runtimeMinutes = 52u,
            tmdbRating = 88u,
            fileId = null,
            seasonCount = seasonCount,
            episodeCount = episodeCount,
        )

    public fun episode(
        id: String = "episode-1",
        number: UInt = 1u,
        title: String = "Pilot",
        fileId: String? = "file-1",
        durationSecs: Double? = 3_120.0,
    ): EpisodeSummary =
        EpisodeSummary(
            id = id,
            episodeNumber = number,
            title = title,
            description = "The handler returns from six years under.",
            thumbnailUrl = "https://beam.test/still.jpg",
            airDate = "2015-05-05",
            durationSecs = durationSecs,
            fileId = fileId,
        )

    public fun season(
        number: UInt = 1u,
        episodes: List<EpisodeSummary> = listOf(episode()),
    ): SeasonSummary =
        SeasonSummary(
            seasonNumber = number,
            posterUrl = null,
            episodeRuntimeMinutes = 52u,
            genres = listOf("Thriller"),
            episodes = episodes,
        )

    public fun movieDetail(summary: MediaSummary = movie()): MediaDetail = MediaDetail.Movie(summary)

    public fun showDetail(
        summary: MediaSummary = show(),
        seasons: List<SeasonSummary> = listOf(season()),
    ): MediaDetail = MediaDetail.Show(summary, seasons)

    public fun page(
        items: List<MediaSummary> = listOf(movie(), show()),
        endCursor: String? = null,
        hasNextPage: Boolean = false,
    ): MediaPage = MediaPage(items, endCursor, hasNextPage)

    public fun source(
        fileId: String = "file-1",
        container: String? = "mkv",
        videoCodec: String? = "hevc",
        width: UInt? = 3840u,
        height: UInt? = 2160u,
        bitRate: ULong? = 24_000_000uL,
        hdrFormat: String? = null,
        sizeBytes: ULong = 18_000_000_000uL,
        audioTracks: List<AudioTrackView> = listOf(audioTrack()),
    ): MediaSourceView =
        MediaSourceView(
            fileId = fileId,
            sizeBytes = sizeBytes,
            durationSecs = 6_300.0,
            container = container,
            mimeType = "video/x-matroska",
            videoCodec = videoCodec,
            width = width,
            height = height,
            bitRate = bitRate,
            hdrFormat = hdrFormat,
            audioTracks = audioTracks,
            streamUrl = "https://beam.test/v1/files/$fileId/stream",
            downloadUrl = "https://beam.test/v1/files/$fileId/download",
        )

    public fun audioTrack(
        codec: String = "eac3",
        language: String? = "eng",
        channels: UShort = 6u,
        isDefault: Boolean = true,
    ): AudioTrackView = AudioTrackView(codec, language, channels, isDefault)

    public fun continueWatching(
        mediaId: String = "movie-1",
        fileId: String = "file-1",
        positionSecs: Double = 1_800.0,
        durationSecs: Double? = 6_300.0,
        media: MediaSummary? = movie(),
        episode: EpisodeSummary? = null,
    ): ContinueWatchingEntry =
        ContinueWatchingEntry(
            mediaId = mediaId,
            episodeId = episode?.id,
            fileId = fileId,
            kind = if (episode == null) MediaKind.MOVIE else MediaKind.SHOW,
            positionSecs = positionSecs,
            durationSecs = durationSecs,
            progressFraction = durationSecs?.let { (positionSecs / it).coerceIn(0.0, 1.0) },
            updatedAtUnix = 1_767_225_600L,
            media = media,
            episode = episode,
        )

    public fun historyEntry(
        mediaId: String = "movie-1",
        completed: Boolean = true,
        media: MediaSummary? = movie(),
    ): HistoryEntry =
        HistoryEntry(
            mediaId = mediaId,
            episodeId = null,
            fileId = "file-1",
            kind = MediaKind.MOVIE,
            positionSecs = 6_300.0,
            durationSecs = 6_300.0,
            progressFraction = 1.0,
            completed = completed,
            updatedAtUnix = 1_767_225_600L,
            media = media,
            episode = null,
        )

    public fun library(
        id: String = "library-1",
        name: String = "Films",
        size: UInt = 412u,
        scanRunning: Boolean = false,
    ): LibrarySummary =
        LibrarySummary(
            id = id,
            name = name,
            description = "Everything on the NAS.",
            size = size,
            lastScanFileCount = 412u,
            lastScanStartedAtUnix = 1_767_225_600L,
            lastScanFinishedAtUnix = if (scanRunning) null else 1_767_225_900L,
        )

    public fun server(
        id: String = "https-beam-local-8000",
        displayName: String = "Home",
        isActive: Boolean = true,
        state: SessionState = SessionState.Authenticated(user()),
    ): ServerSummary =
        ServerSummary(
            id = id,
            displayName = displayName,
            baseUrl = "https://beam.local:8000",
            state = state,
            isActive = isActive,
        )

    public fun user(
        id: String = "user-1",
        displayName: String = "Ada",
        isAdmin: Boolean = true,
    ): UserSummary =
        UserSummary(
            id = id,
            displayName = displayName,
            email = "ada@beam.test",
            isAdmin = isAdmin,
            avatarUrl = null,
        )

    public fun adminStatus(): AdminStatus =
        AdminStatus(
            version = "0.1.0",
            uptimeSecs = 86_400uL,
            counts = AdminCounts(libraries = 3uL, files = 1_204uL, users = 4uL),
            enrichment =
                EnrichmentCounts(
                    pending = 12uL,
                    enriched = 1_180uL,
                    failed = 2uL,
                    unmatched = 10uL,
                ),
            recentScans = emptyList(),
        )
}

File: beam-apple/Packages/BeamKit/Sources/BeamTesting/Fixtures.swift
import BeamCore
import BeamFFI
import BeamModel
import Foundation

/// Builders for the core's records.
///
/// Builders rather than struct literals scattered through the tests, for the
/// reason `beam-domain`'s test-utils exist: a generated record gains a field
/// and every literal in the suite stops compiling, where a builder gains one
/// default. Mirrors `core/testing/Fixtures.kt`.
public enum Fixtures {
    /// A movie in the catalogue.
    public static func movie(
        id: String = "movie-1",
        title: String = "The Third Man",
        year: UInt32? = 1949,
        posterUrl: String? = "https://example.invalid/poster.jpg"
    ) -> MediaSummary {
        MediaSummary(
            id: id,
            kind: .movie,
            title: title,
            originalTitle: title,
            year: year,
            description: "A writer arrives in postwar Vienna.",
            posterUrl: posterUrl,
            backdropUrl: "https://example.invalid/backdrop.jpg",
            genres: ["Film-Noir", "Mystery"],
            runtimeMinutes: 104,
            tmdbRating: 82,
            fileId: "file-\(id)",
            seasonCount: 0,
            episodeCount: 0
        )
    }

    /// A show in the catalogue.
    public static func show(
        id: String = "show-1",
        title: String = "Pole to Pole",
        seasonCount: UInt32 = 2,
        episodeCount: UInt32 = 16
    ) -> MediaSummary {
        MediaSummary(
            id: id,
            kind: .show,
            title: title,
            originalTitle: title,
            year: 1992,
            description: "A journey down the meridian.",
            posterUrl: "https://example.invalid/show.jpg",
            backdropUrl: nil,
            genres: ["Documentary"],
            runtimeMinutes: nil,
            tmdbRating: 88,
            fileId: nil,
            seasonCount: seasonCount,
            episodeCount: episodeCount
        )
    }

    /// One page of catalogue results.
    public static func page(_ items: [MediaSummary], hasNextPage: Bool = false) -> MediaPage {
        MediaPage(
            items: items,
            endCursor: items.last?.id,
            hasNextPage: hasNextPage
        )
    }

    /// A playable source.
    public static func source(
        fileId: String = "file-1",
        container: String? = "mkv",
        videoCodec: String? = "h264",
        width: UInt32? = 1920,
        height: UInt32? = 1080,
        sizeBytes: UInt64 = 4_000_000_000
    ) -> MediaSourceView {
        MediaSourceView(
            fileId: fileId,
            sizeBytes: sizeBytes,
            durationSecs: 6240,
            container: container,
            mimeType: "video/x-matroska",
            videoCodec: videoCodec,
            width: width,
            height: height,
            bitRate: 8_000_000,
            hdrFormat: nil,
            audioTracks: [
                AudioTrackView(codec: "aac", language: "eng", channels: 6, isDefault: true)
            ],
            streamUrl: "https://beam.invalid/v1/files/\(fileId)/stream",
            downloadUrl: "https://beam.invalid/v1/files/\(fileId)/download"
        )
    }

    /// A selection that resolved to hardware decoding.
    public static func selection(
        source: MediaSourceView? = nil,
        rejected: [RejectedSource] = []
    ) -> SourceSelection {
        SourceSelection(
            source: source ?? Self.source(),
            playability: .hardware,
            audioTrackIndex: 0,
            reason: "Hardware decoding",
            rejected: rejected
        )
    }

    /// A source the device cannot play, with the reason.
    public static func rejection(
        fileId: String = "file-2",
        reason: RejectionReason = .videoCodecUnsupported
    ) -> RejectedSource {
        RejectedSource(
            fileId: fileId,
            reason: reason,
            detail: "This device cannot decode VC-1"
        )
    }

    /// A part-watched title.
    public static func continueWatching(
        mediaId: String = "movie-1",
        positionSecs: Double = 1200,
        durationSecs: Double? = 6240
    ) -> ContinueWatchingEntry {
        ContinueWatchingEntry(
            mediaId: mediaId,
            episodeId: nil,
            fileId: "file-\(mediaId)",
            kind: .movie,
            positionSecs: positionSecs,
            durationSecs: durationSecs,
            progressFraction: durationSecs.map { positionSecs / $0 },
            updatedAtUnix: 1_700_000_000,
            media: movie(id: mediaId),
            episode: nil
        )
    }

    /// A watched title.
    public static func historyEntry(
        mediaId: String = "movie-1",
        completed: Bool = false
    ) -> HistoryEntry {
        HistoryEntry(
            mediaId: mediaId,
            episodeId: nil,
            fileId: "file-\(mediaId)",
            kind: .movie,
            positionSecs: completed ? 6240 : 900,
            durationSecs: 6240,
            progressFraction: completed ? 1 : 900 / 6240,
            completed: completed,
            updatedAtUnix: 1_700_000_000,
            media: movie(id: mediaId),
            episode: nil
        )
    }

    /// A library.
    public static func library(
        id: String = "library-1",
        name: String = "Films",
        size: UInt32 = 412
    ) -> LibrarySummary {
        LibrarySummary(
            id: id,
            name: name,
            description: "Everything on the NAS",
            size: size,
            lastScanFileCount: size,
            lastScanStartedAtUnix: 1_700_000_000,
            lastScanFinishedAtUnix: 1_700_000_600
        )
    }

    /// A registered server, signed in.
    public static func server(
        id: String = "server-1",
        isActive: Bool = true
    ) -> ServerSummary {
        ServerSummary(
            id: id,
            displayName: "Home",
            baseUrl: "https://beam.invalid",
            state: .authenticated(
                user: UserSummary(
                    id: "user-1",
                    displayName: "Viewer",
                    email: "viewer@example.invalid",
                    isAdmin: false,
                    avatarUrl: nil
                )
            ),
            isActive: isActive
        )
    }

    /// The HTTP handoff for a file.
    public static func playbackConfig(fileId: String = "file-1") -> PlaybackHttpConfig {
        PlaybackHttpConfig(
            url: "https://beam.invalid/v1/files/\(fileId)/stream",
            headers: ["Cookie": "beam_session=opaque"],
            trustedFingerprints: [],
            pinnedHost: "beam.invalid"
        )
    }
}
